### PR TITLE
Address multiple test262 Intl failures in ICU

### PR DESCRIPTION
**Pull request by jackhorton**

### CODE REVIEW

**Comment by dilijev** [comment]
Out of curiosity, which tests?

**Comment by jackhorton** [comment]
```
PS F:\intl\test262> .\Test-Intl -All -HostType ch -HostPath F:\chbin\nf-icu\x64_release\ch.exe -OutputFile nf-icu.txt
316 passed 258 failed in 22386.8045 ms
PS F:\intl\test262> .\Test-Intl -All -HostType ch -HostPath F:\chbin\intl\test262\x64_release\ch.exe -OutputFile intl-test262.txt
392 passed 182 failed in 19473.4121 ms
```

the following is the diff:
```
FAIL ./test/intl402/Collator/prototype/compare/name.js
FAIL ./test/intl402/Collator/prototype/compare/name.js
FAIL ./test/intl402/constructors-taint-Object-prototype.js
FAIL ./test/intl402/constructors-taint-Object-prototype.js
FAIL ./test/intl402/constructors-taint-Object-prototype-2.js
FAIL ./test/intl402/constructors-taint-Object-prototype-2.js
FAIL ./test/intl402/Date/prototype/throws-same-exceptions-as-DateTimeFormat.js
FAIL ./test/intl402/Date/prototype/throws-same-exceptions-as-DateTimeFormat.js
FAIL ./test/intl402/DateTimeFormat/prototype/format/name.js
FAIL ./test/intl402/DateTimeFormat/prototype/format/name.js
FAIL ./test/intl402/DateTimeFormat/prototype/format/proleptic-gregorian-calendar.js
FAIL ./test/intl402/DateTimeFormat/prototype/format/proleptic-gregorian-calendar.js
FAIL ./test/intl402/DateTimeFormat/prototype/format/throws-value-non-finite.js
FAIL ./test/intl402/DateTimeFormat/prototype/format/throws-value-non-finite.js
FAIL ./test/intl402/DateTimeFormat/prototype/formatToParts/date-is-infinity-throws.js
FAIL ./test/intl402/DateTimeFormat/prototype/formatToParts/date-is-infinity-throws.js
FAIL ./test/intl402/DateTimeFormat/prototype/formatToParts/date-is-nan-throws.js
FAIL ./test/intl402/DateTimeFormat/prototype/formatToParts/date-is-nan-throws.js
FAIL ./test/intl402/DateTimeFormat/taint-Object-prototype-date-time-components.js
FAIL ./test/intl402/DateTimeFormat/taint-Object-prototype-date-time-components.js
FAIL ./test/intl402/DateTimeFormat/timezone-utc.js
FAIL ./test/intl402/DateTimeFormat/timezone-utc.js
FAIL ./test/intl402/default-locale-is-canonicalized.js
FAIL ./test/intl402/default-locale-is-canonicalized.js
FAIL ./test/intl402/default-locale-is-supported.js
FAIL ./test/intl402/default-locale-is-supported.js
FAIL ./test/intl402/fallback-locales-are-supported.js
FAIL ./test/intl402/fallback-locales-are-supported.js
FAIL ./test/intl402/Number/prototype/toLocaleString/default-options-object-prototype.js
FAIL ./test/intl402/Number/prototype/toLocaleString/default-options-object-prototype.js
FAIL ./test/intl402/NumberFormat/currency-code-well-formed.js
FAIL ./test/intl402/NumberFormat/currency-code-well-formed.js
FAIL ./test/intl402/NumberFormat/default-minimum-singificant-digits.js
FAIL ./test/intl402/NumberFormat/default-minimum-singificant-digits.js
FAIL ./test/intl402/NumberFormat/default-options-object-prototype.js
FAIL ./test/intl402/NumberFormat/default-options-object-prototype.js
FAIL ./test/intl402/NumberFormat/ignore-invalid-unicode-ext-values.js
FAIL ./test/intl402/NumberFormat/ignore-invalid-unicode-ext-values.js
FAIL ./test/intl402/NumberFormat/prototype/format/name.js
FAIL ./test/intl402/NumberFormat/prototype/format/name.js
FAIL ./test/intl402/NumberFormat/prototype/resolvedOptions/basic.js
FAIL ./test/intl402/NumberFormat/prototype/resolvedOptions/basic.js
FAIL ./test/intl402/NumberFormat/significant-digits-options-get-sequence.js
FAIL ./test/intl402/NumberFormat/significant-digits-options-get-sequence.js
FAIL ./test/intl402/NumberFormat/supportedLocalesOf/basic.js
FAIL ./test/intl402/NumberFormat/supportedLocalesOf/basic.js
FAIL ./test/intl402/NumberFormat/taint-Object-prototype.js
FAIL ./test/intl402/NumberFormat/taint-Object-prototype.js
FAIL ./test/intl402/NumberFormat/test-option-currency.js
FAIL ./test/intl402/NumberFormat/test-option-currency.js
FAIL ./test/intl402/NumberFormat/test-option-currencyDisplay.js
FAIL ./test/intl402/NumberFormat/test-option-currencyDisplay.js
FAIL ./test/intl402/NumberFormat/test-option-localeMatcher.js
FAIL ./test/intl402/NumberFormat/test-option-localeMatcher.js
FAIL ./test/intl402/NumberFormat/test-option-style.js
FAIL ./test/intl402/NumberFormat/test-option-style.js
FAIL ./test/intl402/NumberFormat/test-option-useGrouping.js
FAIL ./test/intl402/NumberFormat/test-option-useGrouping.js
FAIL ./test/intl402/NumberFormat/throws-for-currency-style-without-currency-option.js
FAIL ./test/intl402/NumberFormat/throws-for-currency-style-without-currency-option.js
FAIL ./test/intl402/supportedLocalesOf-consistent-with-resolvedOptions.js
FAIL ./test/intl402/supportedLocalesOf-consistent-with-resolvedOptions.js
FAIL ./test/intl402/supportedLocalesOf-default-locale-and-zxx-locale.js
FAIL ./test/intl402/supportedLocalesOf-default-locale-and-zxx-locale.js
FAIL ./test/intl402/supportedLocalesOf-duplicate-elements-removed.js
FAIL ./test/intl402/supportedLocalesOf-duplicate-elements-removed.js
FAIL ./test/intl402/supportedLocalesOf-returned-array-elements-are-frozen.js
FAIL ./test/intl402/supportedLocalesOf-returned-array-elements-are-frozen.js
FAIL ./test/intl402/supportedLocalesOf-taint-Array.js
FAIL ./test/intl402/supportedLocalesOf-taint-Array.js
FAIL ./test/intl402/supportedLocalesOf-taint-Array-2.js
FAIL ./test/intl402/supportedLocalesOf-taint-Array-2.js
FAIL ./test/intl402/supportedLocalesOf-test-option-localeMatcher.js
FAIL ./test/intl402/supportedLocalesOf-test-option-localeMatcher.js
FAIL ./test/intl402/supportedLocalesOf-unicode-extensions-ignored.js
FAIL ./test/intl402/supportedLocalesOf-unicode-extensions-ignored.js
```

**Comment by dilijev** [comment]
Cool. Even though these fixes are covered by test262, can you add a few test cases to capture the correct behavior?

File: lib/Runtime/Library/InJavascript/Intl.js [hunk 1]
```diff
@@ -130,7 +130,8 @@
         hasOwnProperty(o, prop) { return callInstanceFunc(platform.builtInJavascriptObjectEntryHasOwnProperty, o, prop); },
         defineProperty: platform.builtInJavascriptObjectEntryDefineProperty,
         isExtensible: platform.builtInJavascriptObjectEntryIsExtensible,
-        create() { return platform.builtInJavascriptObjectCreate(null); },
+        create(proto = null) { return platform.builtInJavascriptObjectCreate(proto); },
+        setPrototypeOf(target, proto = null) { return platform.builtInSetPrototype(target, proto); },
```
Comment (MSLaguana) [inline lib/Runtime/Library/InJavascript/Intl.js]: I found this to be confusing; I saw an instance of `_.setPrototypeOf` and thought that it was the same as the js-spec `Object.setPrototypeOf` which has a required second argument, but had to look back up here to see that this had a default argument. I'd prefer if this was just a straight pass-through, and call-sites explicitly provide null if required.

Comment (jackhorton) [reply lib/Runtime/Library/InJavascript/Intl.js]: Null is what we want to set the prototype to 99% of the time in Intl.js. I have a few such default arguments in the _ object. I was finding providing a second argument to create a lot of code noise.

Comment (jackhorton) [reply lib/Runtime/Library/InJavascript/Intl.js]: _Interesting side note:_ platform.builtInSetPrototype is not the same function as Object.setPrototypeOf, and I am not entirely clear as to why.

**Comment by jackhorton** [comment]
@dilijev One reason to not add tests for these 262 corrections is because winglob still fails a significant number for them, and we don't expose whether we are using winglob or not to the test harness. I had considered adding WScript.Platform.Internationalization to feature-detect in tests, but I would prefer to stay away from having too many extra test files to avoid the current situation where tests are just sortof thrown all over the place and you never know in which file a test should go.

**Comment by dilijev** [comment]
@jackhorton didn't we add tags require_winglob or require_icu?

File: lib/Runtime/Library/InJavascript/Intl.js (hunk 1 above)
Comment (dilijev) [reply lib/Runtime/Library/InJavascript/Intl.js]: ^ Maybe add a comment to that effect so we don't lose track of that bit of trivia -- future maintainer to update the reasoning?

File: lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp [hunk 2]
```diff
@@ -769,13 +769,35 @@ namespace Js
...
+        // zh_TW is equivalent to zh_Hant_TW, which would return true.
+        // We can work around this by searching for both the locale as requested and
+        // the locale in addition to all of its "likely subtags."
+        // This works in practice because, for instance, unum_open("zh_TW") will actually
+        // use "zh_Hant_TW" (confirmed with unum_getLocaleByType(..., ULOC_VALID_LOCALE))
+        // The code below performs the following mappings:
+        // zh_TW -> zh_Hant_TW
+        // zh_CN -> zh_Hans_CN
+        // sr_RS -> sr_Cyrl_RS
+        // pa_PK -> pa_Arab_PK
```
[the remark this answers is not in the document]
Comment (jackhorton) [reply lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp]: They are not exhaustive -- in my local updates I have added "for example". I can alphabetize them too.

**Comment by jackhorton** [comment]
We do have require_winglob and require_icu -- I suppose I can have two different rlexe entries, one for winglob and one for ICU, and pass in different -args for each to let the test know which asserts to run?

File: lib/Runtime/Library/InJavascript/Intl.js [hunk 3]
```diff
@@ -1017,25 +1060,30 @@
...
+            let cDigits = 0;
+            if (useCurrency) {
+                currency = _.toUpperCase(currency);
+                nf.currency = currency;
+                cDigits = platform.currencyDigits(currency);
+            }
 
-                nf.currencyDisplay = GetOption(options, "currencyDisplay", "string", ["code", "symbol", "name"], "symbol");
-                nf.currencyDisplayToUse = toEnum(NumberFormatCurrencyDisplay, _.toUpperCase(nf.currencyDisplay));
+            let currencyDisplay = GetOption(options, "currencyDisplay", "string", ["code", "symbol", "name"], "symbol");
```
Comment (dilijev) [inline lib/Runtime/Library/InJavascript/Intl.js]: >["code", "symbol", "name"] [](start = 82, length = 26)

worried about tainting on array literals like this? (Assuming that fix is coming later though...)

Comment (jackhorton) [reply lib/Runtime/Library/InJavascript/Intl.js]: Yeah, to get ready for windows we should probably take another once-over on the tainting tests

File: lib/Runtime/Library/InJavascript/Intl.js [hunk 4]
```diff
@@ -3489,12 +3570,12 @@
                         platform.raiseNeedObjectOfType("DateTimeFormat.prototype.format", "DateTimeFormat");
                     }
 
-                    if (isNaN(date) || !isFinite(date)) {
+                    if (date !== undefined && (isNaN(date) || !isFinite(date))) {
```
Comment (dilijev) [inline lib/Runtime/Library/InJavascript/Intl.js]: >(isNaN(date) || !isFinite(date)) [](start = 46, length = 32)

=== !isFinite(date)

**Pull request merged**